Libraries/src/Amazon.Lambda.AspNetCoreServer/Internal/HttpRequestMessageConverter.cs

```diff
@@ -0,0 +1,133 @@
+#if NET8_0_OR_GREATER
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Net.Http;
+using System.Text;
+using System.Text.Json;
+using System.Text.Json.Serialization;
+using System.Threading.Tasks;
+using Amazon.Lambda.APIGatewayEvents;
+using Amazon.Lambda.ApplicationLoadBalancerEvents;
+using Microsoft.AspNetCore.Identity.Data;
+using Microsoft.AspNetCore.WebUtilities;
+using Microsoft.Extensions.Primitives;
+
+namespace Amazon.Lambda.AspNetCoreServer.Internal
+{
+    /// <summary>
+    /// Helper class for converting a <see cref="HttpRequestMessage"/> to a known
+    /// lambda request type like: <see cref="APIGatewayHttpApiV2ProxyRequest"/>,
+    /// <see cref="ApplicationLoadBalancerRequest"/>, or <see cref="APIGatewayProxyRequest"/>.
+    /// <para />
+    /// Object is returned as a serialized string.
+    /// <para />
+    /// This is intended for internal use to support SnapStart initialization.  Not all properties
+    /// may be full set.
+    /// </summary>
+    public class HttpRequestMessageConverter
+    {
+        private static readonly Uri _baseUri = new Uri("http://localhost");
+        
+        internal class DuckRequest
+        {
+            public string Body { get; set; }
+            public Dictionary<string, string> Headers { get; set; } = new();
+            public string HttpMethod { get; set; }
+            public string Path { get; set; }
+            public string RawQuery { get; set; }
+            public Dictionary<string, StringValues> Query { get; set; } = new();
+        }
+
+        public static async Task<TRequest> ConvertToLambdaRequest<TRequest>(HttpRequestMessage request)
+        {
+            if (null == request.RequestUri)
+            {
+                throw new ArgumentException($"{nameof(HttpRequestMessage.RequestUri)} must be set.", nameof(request));
+            }
+
+            if (request.RequestUri.IsAbsoluteUri)
+            {
+                throw new ArgumentException($"{nameof(HttpRequestMessage.RequestUri)} must be relative.", nameof(request));
+            }
+
+            // make request absolut (relative to localhost) otherwise parsing the query will not work
+            request.RequestUri = new Uri(_baseUri, request.RequestUri);
+
+            var duckRequest = new DuckRequest
```

Comment:
Not seeing the point of defining the `DuckRequest` type plus an odd name since we don't do duck typing in .NET. Why not just make this body, headers and the others local variables?

Reply:
converted to local variables